hail/src/main/scala/is/hail/expr/ir/EmitCodeBuilder.scala

```diff
@@ -188,14 +188,20 @@ class EmitCodeBuilder(val emb: EmitMethodBuilder[_], var code: Code[Unit]) exten
   }
 
   private def _invoke[T](callee: EmitMethodBuilder[_], _args: Param*): Value[T] = {
-    val expectedArgs = callee.emitParamTypes
+
+    // Instance methods must supply `this` in first position.
+    val expectedArgs =
+      if (callee.mb.isStatic) callee.emitParamTypes
+      else CodeParamType(callee.ecb.cb.ti) +: callee.emitParamTypes
+
     val args = _args.toArray
+
     if (expectedArgs.size != args.length)
-      throw new RuntimeException(s"invoke ${ callee.mb.methodName }: wrong number of parameters: " +
-        s"expected ${ expectedArgs.size }, found ${ args.length }")
-    val codeArgs = args.indices.flatMap { i =>
-      val arg = args(i)
-      val pt = expectedArgs(i)
+      throw new RuntimeException(s"invoke ${callee.mb.methodName}: wrong number of parameters: " +
+        s"expected ${expectedArgs.size}, found ${args.length}"
+      )
```

Comment:
All the type-checking was done in EmitCodeBuilder. Perhaps some generic version should be in CodeBuilderLike.
I don't think this is essential for this change. We can revist this later for a proper tidy-up (which perhaps would be nice to do at some point).

Reply:
OK, I tried to add an assert. This caused way too many failures. In order to get this to work properly I think we need to track more information about class hierarchy in the `TypeInfo` object.

This code fails:

```scala
val F = FunctionBuilder[Int]("F")
...
val a = cb.newLocal("a", Code.newInstance(F.cb, F.cb.ctor, FastSeq())
cb.invoke(F.mb, a) // typecheck failure
```

`a` has type `LocalRef[AsmFunction0[Int]]` the compiler infers it from the FunctionBilder (type `FunctionBuilder[AsmFunction0[Int]]`).
The `TypeInfo` on `F`, however, is `L_C1F;`, which is not the same as the `TypeInfo` on `a`: `Lis/hail/asm4s/AsmFunction0;`

Worthwhile doing, but I think it may be a larger change to get this to work properly.